### PR TITLE
Only pack NSEC3/NSEC3PARAM salt when it is not empty

### LINKED ISSUE #473: Salt is not packed for NSEC3 records
#392 was meant to stop packing the salt from NSEC3/NSEC3PARAM records if the salt was empty (i.e. had the `-` value) but instead just completely stopped packing salts. This causes signature verification failure for signatures of NSEC3/NSEC3PARAM records.

### PR DESCRIPTION
**Pull request by rolandshoemaker**
Instead of just not doing it at all.

Fixes #473.

### CODE REVIEW

**Comment by andrewtj** [comment]
This change won't survive the next run of `go generate` as `zmsg.go` is produced by `msg_generate.go`. I've spiked a fix in #426.

**Comment by rolandshoemaker** [comment]
Realized I was editing a generated file and have pushed the relevant `msg_generate.go` fixes and the generated code.

Just saw your comments on #426, happy to pull this and defer to your changeset if that works better for you guys. Would be very good to get this fixed ASAP though as this is breaking in the wild DNSSEC deployments which is always a super fun thing to debug.

File: msg_generate.go [hunk 1]
```diff
@@ -139,8 +139,10 @@ return off, err
 			case st.Tag(i) == `dns:"base64"`:
 				o("off, err = packStringBase64(rr.%s, msg, off)\n")
 
-			case strings.HasPrefix(st.Tag(i), `dns:"size-hex:SaltLength`): // Hack to fix empty salt length for NSEC3
-				o("if rr.%s == \"-\" { /* do nothing, empty salt */ }\n")
+			case strings.HasPrefix(st.Tag(i), `dns:"size-hex:SaltLength`):
+				// directly write instead of using o() so we get the error check in the correct place
+				field := st.Field(i).Name()
+				fmt.Fprintf(b, "// Only pack salt if value is not \"-\", i.e. empty\nif rr.%s != \"-\" {\noff, err = packStringHex(rr.%s, msg, off)\nif err != nil {\nreturn off, err\n}\n}\n", field, field)
```
Comment (miekg) [inline msg_generate.go]: Please use a string literal here to make it more readable.

**Comment by miekg** [comment]
This can go in before #426

**Pull request merged**